controller/webhook/server.go

```diff
@@ -0,0 +1,152 @@
+package webhook
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/linkerd/linkerd2/controller/k8s"
+	pkgTls "github.com/linkerd/linkerd2/pkg/tls"
+	log "github.com/sirupsen/logrus"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/yaml"
+)
+
+type handlerFunc func(*k8s.API, *admissionv1beta1.AdmissionRequest) (*admissionv1beta1.AdmissionResponse, error)
+
+// Server describes the https server implementing the webhook
+type Server struct {
+	*http.Server
+	api                 *k8s.API
+	handler             handlerFunc
+	controllerNamespace string
+}
+
+// NewServer returns a new instance of Server
+func NewServer(api *k8s.API, addr, name, controllerNamespace string, rootCA *pkgTls.CA, handler handlerFunc) (*Server, error) {
+	c, err := tlsConfig(rootCA, name, controllerNamespace)
+	if err != nil {
+		return nil, err
+	}
+
+	server := &http.Server{
+		Addr:      addr,
+		TLSConfig: c,
+	}
+
+	s := &Server{server, api, handler, controllerNamespace}
+	s.Handler = http.HandlerFunc(s.serve)
+	return s, nil
+}
+
+// Start starts the https server
+func (s *Server) Start() {
+	log.Infof("listening at %s", s.Server.Addr)
+	if err := s.ListenAndServeTLS("", ""); err != nil {
+		if err == http.ErrServerClosed {
+			return
+		}
+		log.Fatal(err)
+	}
+}
+
+func (s *Server) serve(res http.ResponseWriter, req *http.Request) {
+	var (
+		data []byte
+		err  error
+	)
+	if req.Body != nil {
+		data, err = ioutil.ReadAll(req.Body)
+		if err != nil {
+			http.Error(res, err.Error(), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	if len(data) == 0 {
+		return
+	}
+
+	response := s.processReq(data)
+	responseJSON, err := json.Marshal(response)
+	if err != nil {
+		http.Error(res, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if _, err := res.Write(responseJSON); err != nil {
+		http.Error(res, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (s *Server) processReq(data []byte) *admissionv1beta1.AdmissionReview {
+	admissionReview, err := decode(data)
+	if err != nil {
+		log.Error("failed to decode data. Reason: ", err)
+		admissionReview.Response = &admissionv1beta1.AdmissionResponse{
+			UID:     admissionReview.Request.UID,
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: err.Error(),
+			},
+		}
+		return admissionReview
+	}
+	log.Infof("received admission review request %s", admissionReview.Request.UID)
```

Comment:
consider logging the namespace/name rather than UID as a more user-friendly message.

curiously when i turned on debug logging, i wasn't seeing the name in the debug output, any idea what's up?
```bash
$ bin/linkerd logs --control-plane-component controller --container sp-validator
...
linkerd linkerd-controller-c689f4fd9-hpqf6 sp-validator time="2019-04-03T01:22:47Z" level=debug msg="admission request: &AdmissionRequest{UID:fd3716f1-55ae-11e9-adbd-025000000001,Kind:linkerd.io/v1alpha1, Kind=ServiceProfile,Resource:linkerd.io/v1alpha1, Resource=serviceprofiles,SubResource:,Name:,Namespace:linkerd,Operation:CREATE,UserInfo:k8s_io_api_authentication_v1.UserInfo{Username:docker-for-desktop,UID:,Groups:[system:masters system:authenticated],Extra:map[string]ExtraValue{},},Object:k8s_io_apimachinery_pkg_runtime.RawExtension{Raw:*[123
```

Reply:
I've found that `Request.Name` is always sent empty by k8s. From their godoc:
```go
// Name is the name of the object as presented in the request.  On a CREATE operation, the client may omit name and
// rely on the server to generate the name.  If that is the case, this method will return the empty string.
// +optional
Name string `json:"name,omitempty" protobuf:"bytes,5,opt,name=name"`
```

Reply:
😮